### PR TITLE
CI: add opt-in aarch64 linux workflow

### PR DESCRIPTION
**Pull request by snadampal**
Triggered by `ciflow/linux-aarch64` and runs only `test_modules`, `test_mkldnn`, `test_mkldnn_fusion` and `test_openmp` as test for now.
TODOS:
 - Enable sscache for fast CI
 - Extend to a more reasonable test coverage

### CODE REVIEW

File: .github/workflows/linux-aarch64.yml
Comment (malfet) [inline .github/workflows/linux-aarch64.yml:4]:
```suggestion
  # For testing purposes, removeme later
  pull_request:
  push:
```

**Comment by snadampal** [comment]
rebased the PR to the latest PyTorch to get tlparse upgrade (to 0.3.7) changes.

File: .ci/docker/ubuntu/Dockerfile
Comment (atalman) [inline .ci/docker/ubuntu/Dockerfile:193]: Can we update this PR to make this conditional ?

**Comment by atalman** [comment]
@snadampal could you please take a look at failing lint jobs ?

File: .ci/docker/ubuntu/Dockerfile
Comment (snadampal) [reply .ci/docker/ubuntu/Dockerfile:193]: yes, looks like there is no conditional COPY in Dockerfile. I'm checking if there any workaround.

Comment (snadampal) [reply .ci/docker/ubuntu/Dockerfile:193]: I think better option is to have the `pytorch/llvm:9.0.1` available for aarch64-linux as well.

**Comment by malfet** [comment]
Had somebody had a chance to figure out what is amiss with sccache?

File: .ci/docker/ubuntu/Dockerfile
Comment (snadampal) [reply .ci/docker/ubuntu/Dockerfile]: I have work aounded the missing llvm source build layer COPY on aarch64-linux.

**Comment by snadampal** [comment]
Hi @malfet , you'd opened this PR for building sccache from sources, right
https://github.com/pytorch/pytorch/pull/121323

**Comment by snadampal** [comment]
all CI except the Lint related passed. 
@atalman , @malfet , can you please take a look and merge it it looks good to you.

**Comment by atalman** [comment]
Looks good. Need to fix lint, Please use something like this:
```
lintrunner init 
lintrunner -a test/run_test.py
....
```

Also need to add concurrency section to workflow:
```
concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref_name }}-${{ github.ref_type == 'branch' && github.sha }}-${{ github.event_name == 'workflow_dispatch' }}-${{ github.event_name == 'schedule' }}
  cancel-in-progress: true
```

File: .github/workflows/linux-aarch64.yml
Comment (atalman) [inline .github/workflows/linux-aarch64.yml:12]:
```suggestion
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.sha }}-${{ github.event_name == 'workflow_dispatch' }} but found ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref_name }}-${{ github.ref_type == 'branch' && github.sha }}-${{ github.event_name == 'workflow_dispatch' }}-${{ github.event_name == 'schedule' }}
```

According to the workflow error this should be the concurrency

**Comment by snadampal** [comment]
of the 2 test failures, one is a flaky one, failing on the trunk too. and the second one seems to failing because the scripts are using the pre-installed torch library and the variable is new in this PR. I think it will pass once the feature is landed and the torch binaries are updated.

`ImportError: cannot import name 'IS_ARM64_LINUX' from 'torch.testing._internal.common_utils' (/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py)`

**Comment by snadampal** [comment]
All CI tests passed!
cc: @atalman

**Review by atalman — APPROVED** [review]

**Comment by snadampal** [comment]
thanks for approving the PR, @atalman.

@jeffdaily , @malfet , @mruberry , could you please review the PR? 
Currently there is no CI support for aarch64-linux platform and hence we are running into multiple regressions every week. Appreciate if you could prioritize review and merging this soon. thank you!

**Comment by malfet** [comment]
@snadampal can you please add build/test time numbers to this PR?
Otherwise lgtm (though we really need to enable sccache for those)

**Review by malfet — APPROVED** [review]

**Comment by snadampal** [comment]
> @snadampal can you please add build/test time numbers to this PR? Otherwise lgtm (though we really need to enable sccache for those)

sure, but first I will work with you/team to upgrade the runner sizes so that the execution times will be more acceptable.

File: .github/workflows/linux-aarch64.yml
Comment (malfet) [inline .github/workflows/linux-aarch64.yml:5]: This should have been removed before landing

File: .ci/docker/requirements-ci.txt
Comment (malfet) [inline .ci/docker/requirements-ci.txt:270]: Please just update lint runner to 0.12.5 across all platforms

File: test/run_test.py
Comment (malfet) [inline test/run_test.py:1308]: This will render `run_test.py` unusable with any options one might run, and can hinder say CUDA-on-ARM bringup

If you want to limit testing to just test_modules.py, just modify test.sh

**Review by malfet — CHANGES_REQUESTED** [review]
If you are updating lint runner please update it for all platforms rather than for one (and if it could not be done for some reason mention it in PR description)
- Do not make changes that would restrict usability for run_test.py for the purposes of CI, as users should be able to run them on their machines without any restrictions specific to CI
- Do not leave the temporary changes in the workflow job (this supposed to be an opt-in workflow but now it runs on every pull, but not on trunk, which is wrong)

And please allow maintainers suggests changes to your PR, otherwise I could have re-landed it myself

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**
